### PR TITLE
Add Distance[Squared] methods to Vector(2/3)[d]

### PR DESCRIPTION
**Pull request by Frassle**
Also fixes some erroneous comment slashes in Vector3[d].cs.

### CODE REVIEW

**Comment by Frassle** [comment]
Build is failing due to #642.

**Comment by Frassle** [comment]
Sigh
```
      ---- Assert.Equal() Failure
      Expected: 810.1326
      Actual:   810.1326
```
Not sure what's up with that.

**Comment by Nihlus** [comment]
@Frassle That's floating-point math in a nutshell. Assert.Equal performs a bit-by-bit equality comparison which isn't going to work for floating point. Use Assert.ApproximatelyEquivalent (our extension for floating point equivalence) instead.

**Comment by Frassle** [comment]
I was using Assert.ApproximatelyEquivalent:
```
    OpenTK.Tests.Vector2+Distance.DistanceSquared(a, b) = (b - a).LengthSquared [FAIL]
      FsCheck.Xunit.PropertyFailedException : 
      Falsifiable, after 21 tests (0 shrinks) (StdGen (1772121225,296344169)):
      Original:
      ((9.8, 14.25), (-8.909091, -7.2))
      
      ---- Assert.Equal() Failure
      Expected: 810.1326
      Actual:   810.1326
      Stack Trace:
        
        ----- Inner Stack Trace -----
          at OpenTK.Tests.Assert.ApproximatelyEquivalent (System.Single a, System.Single b) [0x00021] in <59a2917cfe42ac81a74503837c91a259>:0 
          at OpenTK.Tests.Vector2+Distance.DistanceSquared(a, b) = (b - a).LengthSquared (OpenTK.Vector2 a, OpenTK.Vector2 b) [0x00017] in <59a2917cfe42ac81a74503837c91a259>:0 
          at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
          at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <204f770036d441bb8dfd3daba3550e83>:0 
        --- End of stack trace from previous location where exception was thrown ---
          at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <204f770036d441bb8dfd3daba3550e83>:0 
          at FsCheck.Runner+invokeAndThrowInner@301-1.Invoke (System.Object[] o) [0x00034] in <591e167b753ee004a74503837b161e59>:0 
          at FsCheck.Runner+funValue@307.Invoke (System.Object x) [0x00013] in <591e167b753ee004a74503837b161e59>:0 
          at <StartupCode$FSharp-Core>.$Reflect+Invoke@835-4[T1,T2].Invoke (T1 inp) [0x00001] in <5939249c904cf4daa74503839c243959>:0 
          at FsCheck.Testable.evaluate[a,b] (Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] body, a a) [0x00003] in <591e167b753ee004a74503837b161e59>:0 
```

**Comment by Nihlus** [comment]
Not sure why it would be failing in that case.

**Comment by Frassle** [comment]
No me neither, especially odd that it works fine on my machine and appveyor. I'll try and repro on Linux this week sometime.

**Comment by varon** [comment]
Silly suggestion for the approximately equal function, but it might solve cases like this. If the string representation of two numbers is equal, then they're approximately equal.

**Comment by varon** [comment]
@frassle - any luck here?

**Comment by Frassle** [comment]
No, have been away will take a look next week probably.

**Comment by Frassle** [comment]
Hmm quick check on linux box, seems accuracy is really off sometimes:

    OpenTK.Tests.Vector3+Distance.DistanceSquared(a, b) = (b - a).LengthSquared [FAIL]
      FsCheck.Xunit.PropertyFailedException :
      Falsifiable, after 76 tests (0 shrinks) (StdGen (2100230347,296348728)):
      Original:
      ((-1.045455, -3.285714, -37.15), (-21.66667, -32.36364, 33.14706))
                                              
      ---- Assert.Equal() Failure
      Expected: 6212.437
      Actual:   6212.436

That's close enough that it's probably float error, but still those numbers don't seem extreme it's strange that there's such a difference at 3 digits.

**Comment by subchannel13** [comment]
Ideally you should use relative tolerance instead of absolute when comparing floats. An option to set tolerance in assert would be desirable too since squaring would naturally have higher precision error then say addition. I now nothing of F# but stitching together my C# and Erlang knowledge I see something interesting in your test framework:
> let approxEqDelta a b = MathHelper.ApproximatelyEqual(a,b,BitAccuracy)
I don't completely agree with MathHelper.ApproximatelyEqual implementation, it may have Hamming cliff problems (similar values with very different binary representation like 1111 and 10000), but something along those lines could help. I'll return when I have something more constructive to say :).

**Comment by subchannel13** [comment]
After reading https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/ and rechecking the code MathHelper.ApproximatelyEqual looks sound for "relative tolerance" comparison.

**Comment by Frassle** [comment]
See that's the oddity. I am using ApproximatelyEquivalent.

    [<Property>]
    let ``DistanceSquared(a, b) = (b - a).LengthSquared`` (a : Vector3, b : Vector3) =
        Assert.ApproximatelyEquivalent(Vector3.DistanceSquared(a, b), (b - a).LengthSquared)

**Comment by subchannel13** [comment]
Yes and Assert.ApproximatelyEquivalent ends up using MathHelper.ApproximatelyEquivalent(a, b, EquivalenceTolerance), not MathHelper.ApproximatelyEqual. Tolerance is set to 0.00005f and it doesn't scale with `a` and `b`'s order of magnitude leading to false negative asertions.

**Comment by Frassle** [comment]
**doh!!**
Thanks @subchannel13, I was being blind. Will update that.

**Comment by Frassle** [comment]
So thanks to @subchannel13 again for pointing out I was using the wrong method. Tests are now passing.

File: src/OpenTK/Math/Vector2.cs [hunk 1]
```diff
@@ -515,6 +515,54 @@ public static void Clamp(ref Vector2 vec, ref Vector2 min, ref Vector2 max, out
...
+            return result;
+        }
+
+        /// <summary>
+        /// Compute the euclidean distance between two vectors.
+        /// </summary>
+        /// <param name="vec1">The first vector</param>
+        /// <param name="vec2">The second vector</param>
+        /// <param name="result">The distance</param>
+        public static void Distance(ref Vector2 vec1, ref Vector2 vec2, out float result)
```
Comment (varon) [inline src/OpenTK/Math/Vector2.cs]: This should probably just return. There's no performance benefit from using a void/out value here.

File: src/OpenTK/Math/Vector2.cs [hunk 2]
```diff
@@ -515,6 +515,54 @@ public static void Clamp(ref Vector2 vec, ref Vector2 min, ref Vector2 max, out
...
+            return result;
+        }
+
+        /// <summary>
+        /// Compute the squared euclidean distance between two vectors.
+        /// </summary>
+        /// <param name="vec1">The first vector</param>
+        /// <param name="vec2">The second vector</param>
+        /// <param name="result">The squared distance</param>
+        public static void DistanceSquared(ref Vector2 vec1, ref Vector2 vec2, out float result)
```
Comment (varon) [inline src/OpenTK/Math/Vector2.cs]: Same here

File: src/OpenTK/Math/Vector2d.cs [hunk 3]
```diff
@@ -508,6 +508,54 @@ public static void Clamp(ref Vector2d vec, ref Vector2d min, ref Vector2d max, o
...
+            return result;
+        }
+
+        /// <summary>
+        /// Compute the euclidean distance between two vectors.
+        /// </summary>
+        /// <param name="vec1">The first vector</param>
+        /// <param name="vec2">The second vector</param>
+        /// <param name="result">The distance</param>
+        public static void Distance(ref Vector2d vec1, ref Vector2d vec2, out double result)
```
Comment (varon) [inline src/OpenTK/Math/Vector2d.cs]: etc

**Comment by Frassle** [comment]
Hmm downside of making them return float/double is you now get a cls compliance warning:

    warning CS3006: Overloaded method 'Vector3d.Distance(ref Vector3d, ref Vector3d)' differing only in ref or out, or in array rank, is not CLS-compliant

**Comment by varon** [comment]
For the CLS compliance, we should hold off on this for 3.0.0 and merge for 4.0.0, I think.

**Comment by varon** [comment]
Or we can go back to the out params. Let's actually just do that, as stupid as the API is, at least then we can get this thing merged.

**Comment by Frassle** [comment]
Ok will flip back, just out of country atm so will do it next week when I'm home.

**Pull request merged**